### PR TITLE
Use a TaskLoop to run container image builds in CI

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Instead of having a CI job per container image, use a TaskLoop
to run one Task per modified container image.
Define a new kaniko task, taken from the catalog, but slighly
modified to add the ability to create the IMAGE parameter
from the context path. This is required because TaskLopps
can only iterate over one parameter.

Note that this CI job will continue to work even if images are
added or removed. If more than one image is modified in a single
PR, they will all be tested, but they will show up on GitHub as
a single CI job.

Signed-off-by: Andrea Frittoli <andrea.frittoli@uk.ibm.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

/kind misc
/hold

### CODE REVIEW

File: tekton/ci-workspace/jobs/tekton-image-build.yaml [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        values: ["passed"]
+      taskRef:
+        apiVersion: custom.tekton.dev/v1alpha1
+        kind: TaskLoop
+        name: image-build
+      params:
+      - name: pullRequestNumber
+        value: "$(params.pullRequestNumber)"
+      - name: context
+        value: $(tasks.check-git-files-changed.results.files)
```
Comment (afrittoli) [inline tekton/ci-workspace/jobs/tekton-image-build.yaml:164]: The new version of `TaskLoop` can take a multiline string as input to the iterate param and break it down into a list.

File: tekton/ci-workspace/jobs/tekton-image-build.yaml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      value: /tekton/home/.docker
+    script: |
+      # Setup the image name
+      CONTEXT=$(workspaces.source.path)/$(params.CONTEXT)
+      IMAGE_NAME=$(basename "${CONTEXT}"):pr-$(params.pullRequestNumber)
+      # Run the build
+      /kaniko/executor \
+        - $(params.EXTRA_ARGS) \
+        --dockerfile=$(params.DOCKERFILE) \
+        --context=$(workspaces.source.path)/$(params.CONTEXT)  # The user does not need to care the workspace and the source.
```
Comment (mpetason) [inline tekton/ci-workspace/jobs/tekton-image-build.yaml:46]:
```suggestion
        --context=$(workspaces.source.path)/$(params.CONTEXT)  # The user does not need to care about the workspace and the source.
```

File: tekton/ci-workspace/jobs/tekton-image-build.yaml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        --destination=${IMAGE_NAME}
+        --digest-file=$(results.IMAGE-DIGEST.path)
+      # kaniko assumes it is running as root, which means this example fails on platforms
+      # that default to run containers as random uid (like OpenShift). Adding this securityContext
```
Comment (mpetason) [inline tekton/ci-workspace/jobs/tekton-image-build.yaml:50]:
```suggestion
      # that default to running containers as random uid (like OpenShift). Adding this securityContext
```

File: tekton/ci-workspace/jobs/tekton-image-build.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  taskRef:
+    name: kaniko4ci
+---
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: tekton-image-build
+  annotations:
+    description: |
+      Build the images of for all the Dockerfile modified by the PR
```
Comment (mpetason) [inline tekton/ci-workspace/jobs/tekton-image-build.yaml:70]:
```suggestion
      Build the images for all the Dockerfiles modified by the PR
```
Maybe build the images of all the Dockerfiles modified by the PR. Of or For

File: tekton/ci-workspace/shared/common-tasks.yaml [hunk 5]
```diff
@@ -5,7 +5,9 @@ metadata:
   annotations:
     description: |
       Returns "passed" in the "check" result if any of the files changed
-      in the source on the workspace matches the regular expression
+      in the source on the workspace matches the regular expression.
+      Returns a "files" result a list of files modified by the PR that
```
Comment (mpetason) [inline tekton/ci-workspace/shared/common-tasks.yaml:9]:
```suggestion
      Returns a "files" result as a list of files modified by the PR that
```

File: tekton/ci-workspace/jobs/tekton-image-build.yaml (hunk 2 above)
Comment (afrittoli) [reply tekton/ci-workspace/jobs/tekton-image-build.yaml:46]: Fixed

File: tekton/ci-workspace/jobs/tekton-image-build.yaml (hunk 3 above)
Comment (afrittoli) [reply tekton/ci-workspace/jobs/tekton-image-build.yaml:50]: Fixed

File: tekton/ci-workspace/jobs/tekton-image-build.yaml (hunk 4 above)
Comment (afrittoli) [reply tekton/ci-workspace/jobs/tekton-image-build.yaml:70]: Fixed

File: tekton/ci-workspace/shared/common-tasks.yaml (hunk 5 above)
Comment (afrittoli) [reply tekton/ci-workspace/shared/common-tasks.yaml:9]: Fixed

**Review by vdemeester — APPROVED** [review]

**Pull request merged**